Add SNMP trap configuration parameters (#449)
infrawatch/documentation: doc-Service-Telemetry-Framework/modules/con_snmp-traps.adoc

[id="snmp-traps_{context}"]
= Sending alerts as SNMP traps

[role="_abstract"]
You can integrate {Project} ({ProjectShort}) with an existing infrastructure monitoring platform that receives notifications through SNMP traps. To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters. SNMP traps are sent using version 2c.

[id="configuration-parameters-for-snmptraps_{context}"]
== Configuration parameters for snmpTraps

> 

Reviewer: Usually when we mention editing the stf object we have an example of the relevant section

Author: I could add an overview here, but since this is a concept module not procedure module, I'm not sure how far down the path I should go.

Author: I looked at this again, and decided this does make sense, and have added an example manifest definition.

[id="snmp-traps_{context}"]
= Sending alerts as SNMP traps

[role="_abstract"]

> You can integrate {Project} ({ProjectShort}) with an existing infrastructure monitoring platform that receives notifications through SNMP traps. To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters. SNMP traps are sent using version 2c.

Reviewer 2: What's version 2c? Is this defined elsewhere in the doc? I don't think it is? Do we need to give the end user an explaination of what it is?

Author: It's the version of the SNMP messages that are being used. If you're using SNMP I think this should be considered prerequisite knowledge. There is no other version available.

Reviewer 2: Thanks Leif, let's leave it as is so

[id="configuration-parameters-for-snmptraps_{context}"]
== Configuration parameters for snmpTraps

The `snmpTraps` parameter contains the following sub-parameters for configuring the alert receiver:

> enabled:: Whether to enable the SNMP trap alert receiver. Values are `true` or `false`. Default is `false`.

Reviewer 3: We could make this line and the next few lines more user focused or full sentences, e.g:
enabled:: Set the value of this to `true` to enable the SNMP trap alert receiver. The default value is `false`.

Reviewer 2: ```suggestion
enabled:: Set the value of this to true to enable the SNMP trap alert receiver. The default value is false.
```

target:: Target address to send SNMP traps. Value is a string. Default is `192.168.24.254`.
port:: Target port to send SNMP traps. Value is an integer. Default is `162`.
community:: Target community to send SNMP traps to. Value is a string. Default is `public`.
retries:: SNMP trap retry delivery limit. Value is an integer. Default is `5`.
timeout:: SNMP trap delivery timeout defined in seconds. Value is an integer. Default is `1`.
alertOidLabel:: Label name in the alert that defines the OID value to send the SNMP trap as. Value is a string. Default is `oid`.
trapOidPrefix:: SNMP trap OID prefix for variable bindings. Value is a string. Default is `1.3.6.1.4.1.50495.15`.
trapDefaultOid:: SNMP trap OID when no alert OID label has been specified with the alert. Value is a string. Default is `1.3.6.1.4.1.50495.15.1.2.1`.
trapDefaultSeverity:: SNMP trap severity when no alert severity has been set. Value is a string. Defaults to an empty string.

> The `snmpTraps` parameter is part of the `alerting.alertmanager.receivers` definition and is configured in the `ServiceTelemetry` object:

Reviewer 3: Maybe: "Configure the `snmpTraps` parameter as part of the `alerting.alertmanager.receivers` definition in the `ServiceTelemetry` object:"

Reviewer 3: ```suggestion
Configure the `snmpTraps` parameter as part of the `alerting.alertmanager.receivers` definition in the `ServiceTelemetry` object:
```

[source,yaml,options="nowrap"]
----
apiVersion: infra.watch/v1beta1
kind: ServiceTelemetry
metadata:
  name: default
  namespace: service-telemetry
spec:
  alerting:
    alertmanager:
      receivers:
        snmpTraps:
          alertOidLabel: oid
          community: public
          enabled: true
          port: 162
          retries: 5
          target: 192.168.25.254
          timeout: 1
          trapDefaultOid: 1.3.6.1.4.1.50495.15.1.2.1
          trapDefaultSeverity: ""
          trapOidPrefix: 1.3.6.1.4.1.50495.15
...
----

[id="overview-of-the-mib-definition_{context}"]
== Overview of the MIB definition

Delivery of SNMP traps uses object identifier (OID) value `1.3.6.1.4.1.50495.15.1.2.1` by default. The management information base (MIB) schema is available at https://github.com/infrawatch/prometheus-webhook-snmp/blob/master/PROMETHEUS-ALERT-CEPH-MIB.txt.

> We can work through the OID definition to understand the additional information provided by a delivered SNMP trap triggered by a Prometheus alert rule delivered by Alertmanager. The value `1.3.6.1.4.1` is a global OID defined for private enterprises. The next identifier `50495` is a private enterprise number assigned by IANA for the Ceph organization. The rest of the values are then child OIDs of the parent.

Reviewer 3: Maybe this line "We can work through the OID definition to understand the additional information provided by a delivered SNMP trap triggered by a Prometheus alert rule delivered by Alertmanager." is a bit informal or kind of fluff. Perhaps we can say something like "The OID number is comprised of the following component values:" and could add a bulleted list also if we like.

Reviewer 3: ```suggestion
The OID number is comprised of the following component values:
* The value `1.3.6.1.4.1` is a global OID defined for private enterprises.
*  The next identifier `50495` is a private enterprise number assigned by IANA for the Ceph organization.
* The other values are child OIDs of the parent.
```

15:: prometheus objects
15.1:: prometheus alerts
15.1.2:: prometheus alert traps
15.1.2.1:: prometheus alert trap default

> The prometheus alert trap default is an object made up of several other sub-objects to OID `1.3.6.1.4.1.50495.15` which is defined by the `alerting.alertmanager.receivers.snmpTraps.trapOidPrefix` parameter:

Reviewer 3: "made up" is a phrasal verb which can be problematic for understanding especially for translation. Could just say "made of" or "made from" or "comprised of" maybe

Author: `comprised of` is fine here thanks

<trapOidPrefix>.1.1.1:: alert name
<trapOidPrefix>.1.1.2:: status
<trapOidPrefix>.1.1.3:: severity
<trapOidPrefix>.1.1.4:: instance
<trapOidPrefix>.1.1.5:: job
<trapOidPrefix>.1.1.6:: description
<trapOidPrefix>.1.1.7:: labels
<trapOidPrefix>.1.1.8:: timestamp
<trapOidPrefix>.1.1.9:: rawdata

> We can look at an example output from a simple SNMP trap receiver that outputs the received trap to the console:

Reviewer 3: The following is example output ...

Reviewer 3: ```suggestion
The following is example output from a simple SNMP trap receiver that outputs the received trap to the console:
```

[id="snmp-traps_{context}"]
= Sending alerts as SNMP traps

[role="_abstract"]

> You can integrate {Project} ({ProjectShort}) with an existing infrastructure monitoring platform that receives notifications through SNMP traps. To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters. SNMP traps are sent using version 2c.

Author: ```suggestion
To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters. SNMP traps are sent using version 2c.
```

Author: CC: @JoanneOFlynn2018 @mickogeary this is where it was moved to
